File: pdui_flutter/lib/widgets/pdui_column.dart
import 'package:flutter/material.dart';

import '../core/pdui_expression_resolver.dart';
import '../proto_out/pdui-proto-out.pb.dart';

class PduiColumn {
  PBExpression pbColumn;

  PduiColumn(this.pbColumn);

  Column resolve() {
    List<Widget> children = List.empty(growable: true);

    for (var child in pbColumn.column.children) {
      children.add(PduiExpressionResolver.parse(child));
    }

    return Column(
      key: pbColumn.key != "" ? Key(pbColumn.key) : null,
      children: children,
    );
  }
}

File: pdui-express/lib/widgets/PDUIColumn.ts
import { PDUIWidget } from "../models/PDUIWidget";
import type { PBColumn, PBExpression } from "../proto-out/pdui-proto-out.ts";

export class PDUIColumn extends PDUIWidget {
    children: PDUIWidget[];

    constructor({ key, children }: { key?: string; children: PDUIWidget[] }) {
        super({
            key,
            identity: "column",
            payload: {
                oneofKind: "column",
                column: {
                    children: children.map((child) => child.toPB()),
                },
            },
        });

        this.children = children;
    }

    override toPB(): PBColumn {
        return {
            children: this.children.map((child) => child.toPB()),
        };
    }

    override toExpression(): PBExpression {
        return {
            key: this.key,
            identity: this.identity,
            payload: {
                oneofKind: "column",
                column: {
                    children: this.children.map((child) =>
                        child.toExpression(),
                    ),
                },
            },
        };
    }
}
